### ISSUE TITLE
DecodeFrameNoDelay errors out on all videos.

### ISSUE THREAD

**Issue by BeStillAsBright** [cisco/openh264#3303]
Hi,

So I'm sure this is me using the API wrong, but I've tried multiple videos and none of them decode. I get wild error numbers as the return value for DecodeFrameNoDelay, and I'm extremely confused because there doesn't seem to be much documentation on how to properly use the library (unless I'm missing something... all I've seen are the github pages and the sample programs which I attempted to essentially adapt to what I need to do).

Anyway, my decoding code is below and I can't for the life of me figure out what I'm doing wrong or different than the sample application, other than that I'm writing in C instead of C++. I'm using the first sample video from here to test: https://www.clipcanvas.com/a/video-clip-formats-and-codec-samples (the 4MB watermarked preview file), though I get similar errors no matter what file I use. I'm getting error code 2111035104 as the return value with that file, but it seems "random" based on the file I'm trying.

I'm guessing my error is in detecting the proper frame size, but that's the sample code for it and it does not seem to work. Any suggestions or guidance would be greatly appreciated.

Thanks!

```C
static int
__decode_video(uint8_t * data,
               size_t    size,
               char    * out_prefix)
{
    ISVCDecoder    * decoder   = NULL;
    SDecodingParam   dec_param = {0};
    SBufferInfo      dst_info  = {0};

    uint8_t * yuv_data[3];

    int64_t pos  = 0;
    int64_t frame_count = 0;

    int ret = -1;

    WelsCreateDecoder(&decoder);

    dec_param.sVideoProperty.eVideoBsType = VIDEO_BITSTREAM_AVC;

    (*decoder)->Initialize(decoder, &dec_param);


    for (frame_count = 0; ; frame_count++) {
        int       frame_size  = 0;
        uint8_t * frame_start = NULL;

        int i = 0;

        for (i = 0; pos + i < size; i++) {
            if ((data[pos + i]     == 0 &&
                 data[pos + i + 1] == 0 &&
                 data[pos + i + 2] == 1 &&
                 i > 1)                 ||
                (data[pos + i]     == 0 &&
                 data[pos + i + 1] == 0 &&
                 data[pos + i + 2] == 0 &&
                 data[pos + i + 3] == 1 &&
                 i > 1)                   ) {
                break;
            }
        }
        frame_size = i;

        frame_start = data + pos;

        printf("Decoding %d bytes\n", frame_size);

        ret = (*decoder)->DecodeFrameNoDelay(decoder,
                                             frame_start, 
                                             frame_size,
                                             yuv_data,
                                             &dst_info);

        if (ret != 0) {
            printf("Error: DecodeFrameNoDelay returned %d\n");
        }

        if (dst_info.iBufferStatus == 1) {

            int width  = 0;
            int height = 0;

            width = dst_info.UsrData.sSystemBuffer.iWidth;
            height = dst_info.UsrData.sSystemBuffer.iHeight;
            __save_frame(yuv_data, width, height, out_prefix, frame_count);
        }

        if (pos + frame_size >= size) {
            break;
        }

        pos += frame_size;
    }

    return 0;
}
```

**Comment by huili2** [comment]
Sorry for late reply.
Absolutely your usage is wrong, for either "Initialize()" or "DecodeFrameNoDelay()".
From the API definition, you can clear see that:
 virtual long EXTAPI Initialize (const SDecodingParam* pParam);
  virtual DECODING_STATE EXTAPI DecodeFrameNoDelay (const unsigned char* kpSrc,
      const int kiSrcLen,
      unsigned char** ppDst,
      SBufferInfo* pDstInfo);

In your usage, you called like "(*decoder)->Initialize(decoder, &dec_param);" and "ret = (*decoder)->DecodeFrameNoDelay(decoder,...".
Please remove the "decoder" in the parameters, and strictly follow the API.

**Comment by BeStillAsBright** [comment]
Right, but I'm writing this in pure C, not C++. In openh264/blob/master/codec/api/svc/codec_api.h, there's an `#ifdef __cplusplus` with the class declarations that those prototypes come from and an `#else` with struct/vtable definitions for when you're working in C. The function pointers for the vtable (and the functions themselves) in the C api look like this:

```C
DECODING_STATE (*DecodeFrameNoDelay) (ISVCDecoder*, const unsigned char* pSrc,
                                      const int iSrcLen,
                                      unsigned char** ppDst,
                                      SBufferInfo* pDstInfo);
```
which definitely requires passing the decoder as the first parameter. My code is definitely compiling, so having that as the first argument shouldn't be the issue. I'm wondering if the issue is that my video may be in a container format as opposed to just raw h264, but I'm not sure and can't seem to find sample data to try this with.

Thanks for the reply!

**Comment by huili2** [comment]
Oh sorry for misunderstanding above.
So could you please try to dump some headers of the input "data", to see if it is in pure h264 format?
btw, the error code is quite strange there....
